src/rsync.rs

```diff
@@ -1,76 +1,259 @@
+        trace!("Writing rsync file {:?}", &path);
+        file_ops::write_buf(&path, element.data())?;
+    }
+
+    Ok(())
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+struct RsyncDirState {
+    current: Option<RsyncRevision>,
+    old: Vec<DeprecatedRsyncRevision>,
+}
+
+impl RsyncDirState {
+    /// Gets the current state from disk, if a state file exists. Otherwise returns
+    /// a new blank state.
+    fn recover(config: &Config) -> Result<Self> {
+        let state_path = config.rsync_state_path();
+        if state_path.exists() {
+            let json_bytes = file_ops::read_file(&state_path)
+                .with_context(|| format!("Cannot read rsync state file at: {:?}", state_path))?;
+            serde_json::from_slice(json_bytes.as_ref()).with_context(|| {
+                format!(
+                    "Cannot deserialize json for current state from {:?}",
+                    state_path
+                )
+            })
+        } else {
+            Ok(RsyncDirState {
+                current: None,
+                old: vec![],
+            })
+        }
+    }
+
+    /// Persists the state to disk
+    fn persist(&self, config: &Config) -> Result<()> {
+        let state_path = config.rsync_state_path();
+        let json = serde_json::to_string_pretty(&self)?;
+        file_ops::write_buf(&state_path, json.as_bytes()).with_context(|| "Could not save state.")
+    }
+
+    /// Updates the current revision for this state, moves a possible
+    /// existing current state to old.
+    fn update_current(&mut self, current: RsyncRevision) {
+        let existing = std::mem::replace(&mut self.current, Some(current));
+        if let Some(existing) = existing {
+            self.old.push(existing.deprecate());
+        }
+    }
+
+    /// Cleans old directories from disk when their time has come, and updates
+    /// this state (forgets these old versions). Will throw an error if removing
+    /// an old dir fails, but will simply skip removing old dirs if they had
+    /// already been removed.
+    fn clean_old(&mut self, config: &Config) -> Result<()> {
+        let clean_before = Time::seconds_ago(config.cleanup_after);
+
+        for old in self
+            .old
+            .iter()
+            .filter(|deprecated| deprecated.since < clean_before)
+        {
+            let path = old.revision.path(config);
+            if path.exists() {
+                info!(
+                    "Removing rsync directory: {:?}, deprecated since: {}",
+                    path, old.since
+                );
+                // Try to remove the old directory if it still exists
+                std::fs::remove_dir_all(&path).with_context(|| {
+                    format!("Could not remove rsync dir for old revision at: {:?}", path)
+                })?;
+            }
+        }
+
+        self.old
+            .retain(|deprecated| deprecated.since > clean_before);
```

Comment:
Question: Above the filter uses <, and here retain uses >, so what about the == case? I'm in a bit of a rush so I haven't thought this through, just wondering if there's a gap that something fell through here.

Reply:
Thanks! Yes, this would have been a race condition.